Add safety checks when rendering kernel key strings

dbort opened the pull request:
Summary:
The old code assumed that it was handed a MAX_SIZE buffer, and that the list of TensorMeta values would never generate a string longer than that size.

This PR adds explicit size tracking and an error code to the API, and now returns an error if the buffer is too small for the provided values.

While I'm here, move MAX_SIZE out of the public API, since it's not an intrinsic aspect of kernel keys. This is technically a BC-breaking change, but I don't expect that any users are actually depending on it.

Also:
 - Fix some unused var warnings, which uncovered a couple places where we ignored an Error. Along similar lines, replaced EXPECT with ASSERT when validating kernel registration so we don't try to run the rest of the test after a registration error.
- Silence the Meta-internal linter by moving to std::array in several code chunks I needed to touch anyway because of the new `make_kernel_key` parameter.

Add unit tests for all modified code.

Differential Revision: D69324821

larryliu0820 commented on runtime/kernel/operator_registry.cpp:
```diff
@@ -114,44 +114,106 @@ Error register_kernels(const Span<const Kernel> kernels) {
+
+  // Reserve one byte for null terminator.
+  if (buf_size < 1) {
+    return Error::InvalidArgument;
```
Do you think it's useful to have a new error type like BufferTooSmall?

dbort replied:
I think that's too specific, and not very actionable by users who get it. Even InvalidArgument isn't very useful once it's bubbled up to higher levels tbh. The logs are what will provide the most information.

larryliu0820 reviewed (approved):
Thank you for adding the check! Just left one comment

dbort committed:
Add safety checks when rendering kernel key strings (#8327)

Summary:

The old code assumed that it was handed a MAX_SIZE buffer, and that the list of TensorMeta values would never generate a string longer than that size.

This PR adds explicit size tracking and an error code to the API, and now returns an error if the buffer is too small for the provided values.

While I'm here, move MAX_SIZE out of the public API, since it's not an intrinsic aspect of kernel keys. This is technically a BC-breaking change, but I don't expect that any users are actually depending on it.

Also:
 - Fix some unused var warnings, which uncovered a couple places where we ignored an Error. Along similar lines, replaced EXPECT with ASSERT when validating kernel registration so we don't try to run the rest of the test after a registration error.
- Silence the Meta-internal linter by moving to std::array in several code chunks I needed to touch anyway because of the new `make_kernel_key` parameter.

Add unit tests for all modified code.

Reviewed By: larryliu0820

Differential Revision: D69324821
